### ISSUE TITLE
Streaming output not the same as input

### ISSUE THREAD

**Issue by ZerGo0** [httptoolkit/brotli-wasm#22]
Hey, I'm currently trying to use the streaming example, but for some reason it doesn't work as expected. The output itself is correct, but it's too short.

This is the code (Pretty much the example, but with a file as input):
```
  async Test(file: UploadFile) {
    const fileStream = file.stream();

    const zipper = await brotliPromise;
    const compressStream = new zipper.CompressStream();
    const compressionStream = new TransformStream({
      start() { },
      transform(chunk, controller) {
        controller.enqueue(compressStream.compress(chunk, 100));
      },
      flush(controller) {
        if (compressStream.result() === 2) {
          controller.enqueue(compressStream.compress(undefined, 100));
        }
        controller.terminate();
      }
    });

    const decompressStream = new zipper.DecompressStream();
    const decompressionStream = new TransformStream({
      start() { },
      transform(chunk, controller) {
        controller.enqueue(decompressStream.decompress(chunk, 100));
      },
      flush(controller) {
        if (decompressStream.result() === 2) {
          controller.enqueue(decompressStream.decompress(undefined, 100));
        }
        controller.terminate();
      }
    });

    const textDecoderStream = new TextDecoderStream()
    const outputStream = new WritableStream({
      write(chunk) {
        output += chunk;
      }
    });

    let output = '';
    await fileStream
      .pipeThrough(compressionStream)
      .pipeThrough(decompressionStream)
      .pipeThrough(textDecoderStream)
      .pipeTo(outputStream);
    console.log('File Size: ', file.size);
    console.log('Output Size: ', output.length);
  }
```
Here is the test file: [test.txt](https://github.com/httptoolkit/brotli-wasm/files/10787640/test.txt)

I'm doing this in a browser, Firefox 111.0 to be exact. The filestream returns the expected output, but the compression stream returns a empty array most of the time.

I use vite and excluded the wasm file from optimizations like this:
```
  optimizeDeps: {
    exclude: ["brotli-wasm", "brotli-wasm/pkg.bundler/brotli_wasm_bg.wasm"],
  },
```

Any idea what I could be doing wrong?

**Comment by ZerGo0** [comment]
> Not sure I'm afraid, I don't know what could cause this. Streaming support was added in #11, @myl7 any ideas what could be breaking in this case?
> 
> If you clone this repo, and try doing the same thing inside the test suite here, do you see the same result? That would be helpful to exclude your project config (and give us a conveniently failing test we can try to fix, assuming it does fail).

I'm only able to run the node tests for reason, but they don't fail. However when I modify one of the tests so that the input is a couple of megabytes it fails:

```
it("can streamingly compress & decompress back to the original result (LONG)", () => {
    let s = "";
    for (let i = 0; i < 100000; i++) {
      s += "Brotli brotli brotli brotli";
    }

    const encInput = textEncoder.encode(s);
    const encInput1 = encInput.slice(0, encInput.length / 2);
    const encInput2 = encInput.slice(encInput.length / 2);
    const encStream = new brotli.CompressStream();
    const encOutput1 = encStream.compress(encInput1, 100);
    expect(encStream.result()).to.equal(brotli.BrotliStreamResult.NeedsMoreInput);
    const encOutput2 = encStream.compress(encInput2, 100);
    expect(encStream.result()).to.equal(brotli.BrotliStreamResult.NeedsMoreInput);
    const encOutput3 = encStream.compress(undefined, 100);
    expect(encStream.result()).to.equal(brotli.BrotliStreamResult.ResultSuccess);
    const encOutput = new Uint8Array([...encOutput1, ...encOutput2, ...encOutput3]);

    const decInput1 = encOutput.slice(0, encOutput.length / 2);
    const decInput2 = encOutput.slice(encOutput.length / 2);
    const decStream = new brotli.DecompressStream();
    const decOutput1 = decStream.decompress(decInput1, 100);
    expect(decStream.result()).to.equal(brotli.BrotliStreamResult.NeedsMoreInput);
    const decOutput2 = decStream.decompress(decInput2, 100);
    expect(decStream.result()).to.equal(brotli.BrotliStreamResult.ResultSuccess);
    const decOutput = new Uint8Array([...decOutput1, ...decOutput2]);

    expect(textDecoder.decode(decOutput)).to.equal(s);
  });
```

**Comment by ZerGo0** [comment]
> Not sure I'm afraid, I don't know what could cause this. Streaming support was added in #11, @myl7 any ideas what could be breaking in this case?
> 
> If you clone this repo, and try doing the same thing inside the test suite here, do you see the same result? That would be helpful to exclude your project config (and give us a conveniently failing test we can try to fix, assuming it does fail).

I'm only able to run the node tests and none of them fail, however when I create a test that has a longer input and uses the example code it results in the same issue that I'm having. Decompression seems to stop after the first chunk pretty much.

```
  it("can streamingly compress & decompress back to the original result (Long version)", async () => {
    let input = "";
    for (let i = 0; i < 10; i++) {
      input += "Brotli brotli brotli brotli";
    }

    const inputStream = new ReadableStream({
      start(controller) {
        controller.enqueue(input);
        controller.close();
      }
    });

    const textEncoderStream = new TextEncoderStream();
    const compressStream = new brotli.CompressStream();
    const compressionStream = new TransformStream({
      start() { },
      transform(chunk, controller) {
        controller.enqueue(compressStream.compress(chunk, 100));
      },
      flush(controller) {
        if (compressStream.result() === brotli.BrotliStreamResult.NeedsMoreInput) {
          controller.enqueue(compressStream.compress(undefined, 100));
        }
        controller.terminate();
      }
    });

    const decompressStream = new brotli.DecompressStream();
    const decompressionStream = new TransformStream({
      start() { },
      transform(chunk, controller) {
        controller.enqueue(decompressStream.decompress(chunk, 100));
      },
      flush(controller) {
        if (decompressStream.result() === brotli.BrotliStreamResult.NeedsMoreInput) {
          controller.enqueue(decompressStream.decompress(undefined, 100));
        }
        controller.terminate();
      }
    });

    const textDecoderStream = new TextDecoderStream()
    const outputStream = new WritableStream({
      write(chunk) {
        output += chunk;
      }
    });

    let output = '';
    await inputStream
      .pipeThrough(textEncoderStream)
      .pipeThrough(compressionStream)
      .pipeThrough(decompressionStream)
      .pipeThrough(textDecoderStream)
      .pipeTo(outputStream);

    expect(input.length).to.equal(output.length);
  });
```

**Comment by pimterry** [comment]
@ZerGo0 I think I've work out the correct way that this should have been working. The core issue was that the `NeedsMoreInput` check for decompression should have been `NeedsMoreOutput`, but there were a couple of other issues here too. None of these affect small toy examples, but they did break as soon as you tried to use web streams with any more substantial chunks of data.

I've now updated the [README example](https://github.com/httptoolkit/brotli-wasm#in-browser-with-streams) so it works fully, and there's a full (passing) test in the test suite here: https://github.com/httptoolkit/brotli-wasm/blob/b5c095da32a2c646b6e8c1771d83f437c1777e35/test/brotli.spec.ts#L283-L352

If you change your setup to use that, does it now work for you?

I'm still not sure this is perfect in all cases (I'm not totally sure how some of the missing input/output buffering works internally) but it seems to now work correctly in every case I've tested so it should be very significantly more robust.
